### PR TITLE
edge: drm: zocl: Fixes in Partial Reconfiguration feature support

### PR DESCRIPTION
**Pull request by rajkumar-xilinx**
Fixed pcap download code in zocl driver for Partial reconfiguration support on edge shell.
xclbin downloaded successfully through pcap with these fixes.

### CODE REVIEW

**Comment by rajkumar-xilinx** [comment]
I will fix checkpatch warnings and will update.  Currently, I am seeing below error. Once this issue resolved, I will update the patch.
fatal: unable to access 'https://github.com/Xilinx/XRT.git/': Failed connect to proxy:8080; Connection refused

File: src/runtime_src/core/edge/user/shim.cpp [hunk 1]
```diff
@@ -462,8 +462,12 @@ int ZYNQShim::xclLoadAxlf(const axlf *buffer)
 	}
 
 #if defined(XCLBIN_DOWNLOAD)
-  drm_zocl_pcap_download obj = { const_cast<axlf *>(buffer) };
-  ret = ioctl(mKernelFD, DRM_IOCTL_ZOCL_PCAP_DOWNLOAD, &obj);
+	drm_zocl_pcap_download obj = { const_cast<axlf *>(buffer) };
+	ret = ioctl(mKernelFD, DRM_IOCTL_ZOCL_PCAP_DOWNLOAD, &obj);
+	if (0 != ret)
```
Comment (mamin506) [inline src/runtime_src/core/edge/user/shim.cpp]: Use this style like the other error check in shim.cpp.
    if (!ret)

File: src/runtime_src/core/edge/user/shim.cpp [hunk 2]
```diff
@@ -462,8 +462,12 @@ int ZYNQShim::xclLoadAxlf(const axlf *buffer)
 	}
 
 #if defined(XCLBIN_DOWNLOAD)
-  drm_zocl_pcap_download obj = { const_cast<axlf *>(buffer) };
-  ret = ioctl(mKernelFD, DRM_IOCTL_ZOCL_PCAP_DOWNLOAD, &obj);
+	drm_zocl_pcap_download obj = { const_cast<axlf *>(buffer) };
+	ret = ioctl(mKernelFD, DRM_IOCTL_ZOCL_PCAP_DOWNLOAD, &obj);
```
Comment (mamin506) [inline src/runtime_src/core/edge/user/shim.cpp]: Why we still need DRM_IOCTL_ZOCL_PCAP_DOWNLOAD?
Download bitstream in DRM_IOCTL_ZOCL_READ_AXLF, if BITSTREAM section is found.

File: src/runtime_src/core/edge/user/shim.cpp [hunk 3]
```diff
@@ -462,8 +462,12 @@ int ZYNQShim::xclLoadAxlf(const axlf *buffer)
...
+	if (0 != ret)
+		std::cout << __func__ << "PCAP download failed, err: " << ret << std::endl;
+	else
+		std::cout << __func__ << "PCAP download successful, ret: " << ret << std::endl;
```
Comment (mamin506) [inline src/runtime_src/core/edge/user/shim.cpp]: Do not print on screen unless it is an error.

File: src/runtime_src/core/edge/drm/zocl/zocl_drv.h [hunk 4]
```diff
@@ -97,6 +97,9 @@
 	#define ZOCL_ACCESS_OK(TYPE, ADDR, SIZE) access_ok(TYPE, ADDR, SIZE)
 #endif
 
+/* Enable this macro for partial reconfiguration feature */
+//#define XCLBIN_DOWNLOAD
```
[the remark this answers is not in the document]
Comment (larry9523) [reply src/runtime_src/core/edge/drm/zocl/zocl_drv.h]: I think the behavior should not be controlled by a Macro. The same code binary should work on both PR and non-PR platform

File: src/runtime_src/core/edge/drm/zocl/zocl_ioctl.c [hunk 5]
```diff
@@ -248,9 +257,23 @@ static int zocl_pcap_download(struct drm_zocl_dev *zdev,
 		data[i+1] = data[i+2];
 		data[i+2] = temp;
 	}
-#endif
 
-	err = fpga_mgr_buf_load(fpga_mgr, 0, data, bit_header.BitstreamLength);
+	/* struct with information about the FPGA image to program. */
+	info = fpga_image_info_alloc(&fpga_mgr->dev);
```
Comment (larry9523) [inline src/runtime_src/core/edge/drm/zocl/zocl_ioctl.c]: Should we call fpga_image_info_free() before bailing out?

File: src/runtime_src/core/edge/drm/zocl/zocl_ioctl.c [hunk 6]
```diff
@@ -248,9 +257,23 @@ static int zocl_pcap_download(struct drm_zocl_dev *zdev,
...
+
+	info->buf = data;
+	info->count = bit_header.BitstreamLength;
+	info->sgt = NULL;
+	info->firmware_name = NULL;
+	info->flags = FPGA_MGR_PARTIAL_RECONFIG;
+
+	/* Load the buffer to the FPGA */
+	err = fpga_mgr_load(fpga_mgr, info);
 	DRM_INFO("%s : ret code %d\n", __func__, err);
```
Comment (larry9523) [inline src/runtime_src/core/edge/drm/zocl/zocl_ioctl.c]: The DRM_INFO should be replaced by DRM_ERROR on the non-zero return value of err

File: src/runtime_src/core/edge/drm/zocl/zocl_ioctl.c [hunk 7]
```diff
@@ -248,9 +257,23 @@ static int zocl_pcap_download(struct drm_zocl_dev *zdev,
...
 
 	goto free_buffers;
```
Comment (larry9523) [inline src/runtime_src/core/edge/drm/zocl/zocl_ioctl.c]: this goto is not necessary

**Comment by larry9523** [comment]
I am thinking of the option to make this PR less risky. Can we keep the default behavior, not downloading bitstream, and enable the Partial Reconfigurtaion through xrt.ini (sdaccel.ini)?

**Comment by rajkumar-xilinx** [comment]
Taken care of review comments.
Enabling partial reconfig feature through xrt.ini config file at runtime.
So, removed macro XCLBIN_DOWNLOAD from XRT.
We can enable partial reconfig feature through xrt.ini file now.
Add enable_pr entry under [Runtime] section to enable it.
xrt.ini file contents for example:
[Runtime]
enable_pr=true

Passing pr isolation address through DT.

Validated the changes on zynqMP, Partial reconfig feature is working.

@mamin506 We have kept PCAP_DOWNLOAD icotl as it is for partial reconfig. Since, we are enabling this feature through xrt.ini configuration.

File: src/runtime_src/core/edge/drm/zocl/zocl_drv.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (rajkumar-xilinx) [reply src/runtime_src/core/edge/drm/zocl/zocl_drv.h]: I have removed this macro, and added changes in XRT for using xrt.ini config file to enable/disable this feature.

File: src/runtime_src/core/edge/drm/zocl/zocl_ioctl.c [hunk 8]
```diff
@@ -248,12 +259,30 @@ static int zocl_pcap_download(struct drm_zocl_dev *zdev,
...
+	info->count = bit_header.BitstreamLength;
+	info->sgt = NULL;
+	info->firmware_name = NULL;
+	info->flags = FPGA_MGR_PARTIAL_RECONFIG;
+
+	/* Load the buffer to the FPGA */
+	err = fpga_mgr_load(fpga_mgr, info);
+	if (err) {
+		DRM_ERROR("%s : ret code %d\n", __func__, err);
+		fpga_image_info_free(info);
```
Comment (larry9523) [inline src/runtime_src/core/edge/drm/zocl/zocl_ioctl.c]: Shouldn't we call fpga_image_info_free() no matter we successfully download bitstream or not?

**Comment by chvamshi-xilinx** [comment]
Hi @mamin506  & @larry9523 , 
Taken care of the feedback. Could you please approve the change if everything looks good. 
Thanks

**Comment by maxzhen** [comment]
ok to test

**Comment by chvamshi-xilinx** [comment]
retest this please

**Pull request merged**